### PR TITLE
Model evaluation

### PR DESCRIPTION
**Pull request by lperryp**
## Related Issues
fixes: link to the issue

## Description
- Adding new script for evaluation of ML model 
- Calling the new script from DBotTrainTextClassifier

## Must have
- [ ] Tests
- [ ] Documentation (with link to it)
- [ ] Code Review

## Technical writer review
Mention and link to the files that require a technical writer review.
- [ ] [YAML file](link)
- [ ] [CHANGELOG](link)

### CODE REVIEW

File: Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py [hunk 1]
```diff
@@ -1,14 +1,15 @@
...
 import demisto_ml
 import pandas as pd
-from tabulate import tabulate
-import base64
+from sklearn.model_selection import StratifiedKFold
+
+from CommonServerPython import *
 
+TRAIN_SIZE_FOR_EVALUATION = 0.8
+DEFAULT_LABEL_PREFIX = '|tag|'
```
Comment (erezh31) [inline Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py]: I don't think we need this here

File: Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py [hunk 2]
```diff
@@ -271,14 +221,30 @@ def main():
...
+    X = pd.Series(train_text_data)
+    y = pd.Series(train_tag_data)
+    n_splits = int(1.0 / (1 - TRAIN_SIZE_FOR_EVALUATION))
+    skf = StratifiedKFold(n_splits=n_splits, shuffle=False, random_state=None)
+    skf.get_n_splits(X, y)
+    train_index, test_index = list(skf.split(X, y))[-1]
+    X_train, X_test = list(X[train_index]), list(X[test_index])
+    y_train, y_test = list(y[train_index]), list(y[test_index])
+    model = demisto_ml.train_text_classifier(X_train, y_train)
+    ft_test_predictions = demisto_ml.predict(model, X_test, DEFAULT_LABEL_PREFIX)
```
Comment (erezh31) [inline Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py]: just call this function without the DEFAULT_LABEL_PREFIX - it has it

File: Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.yml [hunk 3]
```diff
@@ -48,6 +48,7 @@ args:
   name: maxBelowThreshold
   required: false
   secret: false
+  deprecated: true
```
Comment (erezh31) [inline Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.yml]: also, change the default value of this argument to empty - cuz it's deprecated

File: Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.yml [hunk 4]
```diff
@@ -113,6 +114,7 @@ args:
   - weighted avgm
   required: false
   secret: false
```
Comment (erezh31) [inline Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.yml]: change the default value to empty

File: Scripts/EvaluateMLModel/EvaluateMLModel.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+                           '\n']
+    human_readable = []
+    human_readable += human_readable_threshold + ['\n']
+    human_readable += class_metrics_human_readable + ['\n']
+    human_readable += csr_matrix_readable
+    human_readable += csr_matrix_readable
+    human_readable = '\n'.join(human_readable)
+    contents = {'threshold': threshold, 'csr_matrix_at_threshold': csr_matrix_at_threshold.to_json(),
+                'metrics_df': metrics_df.to_string()}
+    entry = {
```
Comment (erezh31) [inline Scripts/EvaluateMLModel/EvaluateMLModel.py]: Add entry context + outputs

File: Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py (hunk 2 above)
Comment (lperryp) [reply Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py]: Right, changed, thanks!

File: Scripts/EvaluateMLModel/EvaluateMLModel.py (hunk 5 above)
Comment (lperryp) [reply Scripts/EvaluateMLModel/EvaluateMLModel.py]: added, thanks

File: Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py [hunk 6]
```diff
@@ -271,14 +221,31 @@ def main():
...
         pass
-
-    # evaluate model
-    human_readable, confusion_matrix, report = evaluate_model(train_text_data,
-                                                              train_tag_data,
-                                                              target_accuracy,
-                                                              max_samples_below_threshold,
-                                                              metric)
-
+    X = pd.Series(train_text_data)
```
Comment (erezh31) [inline Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py]: please wrap this logic in a function

File: Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py [hunk 7]
```diff
@@ -287,15 +254,43 @@ def main():
...
+        'Contents': res[0]['Contents'],
         'ContentsFormat': formats['json'],
         'HumanReadable': human_readable,
         'HumanReadableFormat': formats['markdown'],
         'EntryContext': {
             'DBotPhishingClassifier': {
                 'ModelName': model_name,
-                'EvaluationScores': {k.strip().replace(" ", "_"): v for k, v in report.items()},
-                'ConfusionMatrix': confusion_matrix.to_dict()
+                'EvaluationScores': json.loads(res[0]['Contents']['metrics_df']),
```
Comment (erezh31) [inline Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py]: please set evlation scores as numbers

File: Scripts/GetMLModelEvaluation/GetMLModelEvaluation.py [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+        'higher confidence than threshold',
+        'Total - The number of mails from the class in the evaluation set (above and below threshold)',
+
+    ]
+    df.set_index('Class', inplace=True)
+    return df, explanation
+
+
+def output_report(y_true, y_true_per_class, y_pred, y_pred_per_class, threshold):
+    global DETAILED_OUTPUT
```
Comment (erezh31) [inline Scripts/GetMLModelEvaluation/GetMLModelEvaluation.py]: don't use global here - you can get this value as argument to the function

File: Scripts/GetMLModelEvaluation/GetMLModelEvaluation.py [hunk 9]
```diff
@@ -0,0 +1,200 @@
+import numpy as np
+import pandas as pd
+from sklearn.metrics import precision_score, recall_score
+from tabulate import tabulate
+
+from CommonServerPython import *
+
+# pylint: disable=no-member
+
+DETAILED_OUTPUT = True
```
Comment (erezh31) [inline Scripts/GetMLModelEvaluation/GetMLModelEvaluation.py]: make it demisto.args with default

File: Scripts/GetMLModelEvaluation/GetMLModelEvaluation.py [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+                        'Precision': precision,
+                        'Recall': recall,
+                        'TP': classified_correctly,
+                        'CoverageInt': int(above_thresh),
+                        'Total': total}, ignore_index=True)
+    df = df.append({'Class': 'All',
+                    'Precision': df["Precision"].mean(),
+                    'Recall': df["Recall"].mean(),
+                    'TP': df["TP"].sum(),
+                    'CoverageInt': df["CoverageInt"].sum(),
```
Comment (erezh31) [inline Scripts/GetMLModelEvaluation/GetMLModelEvaluation.py]: Add also FP

File: Scripts/GetMLModelEvaluation/GetMLModelEvaluation.py [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+        recall = recall_score(y_true=y_true_class, y_pred=y_pred_class_binary)
+        classified_correctly = sum(1 for y_true_i, y_pred_i in zip(y_true_class, y_pred_class_binary) if y_true_i == 1
+                                   and y_pred_i == 1)
+        above_thresh = sum(1 for i, y_true_i in enumerate(y_true_class) if y_true_i == 1
+                           and any(y_pred_per_class[c][i] >= threshold for c in y_pred_per_class))
+        total = int(sum(y_true_class))
+        df = df.append({'Class': class_,
+                        'Precision': precision,
+                        'Recall': recall,
+                        'TP': classified_correctly,
```
Comment (erezh31) [inline Scripts/GetMLModelEvaluation/GetMLModelEvaluation.py]: Add also FP

File: TestPlaybooks/playbook-Create_Phishing_Classifier_V2_ML_Test.yml [hunk 12]
```diff
@@ -271,7 +271,7 @@ tasks:
       - - operator: isNotEmpty
           left:
             value:
-              simple: DBotPhishingClassifier.EvaluationScores.micro_avg
+              simple: DBotPhishingClassifier.EvaluationScores.Precision
```
Comment (erezh31) [inline TestPlaybooks/playbook-Create_Phishing_Classifier_V2_ML_Test.yml]: Add the changes I sent you to the playbook outputs

File: Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py (hunk 1 above)
Comment (lperryp) [reply Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py]: changed accordingly

File: Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py [hunk 13]
```diff
@@ -260,25 +209,30 @@ def main():
...
         demisto.results(entry)
-
-    train_tag_data = map(lambda x: x[DBOT_TAG_FIELD], data)
-    train_text_data = map(lambda x: x[text_field], data)
-    if len(train_text_data) != len(train_tag_data):
-        return_error("Error: data and tag data are different length")
-
     # print important words for each category
     try:
         find_keywords(data, DBOT_TAG_FIELD, text_field, keyword_min_score)
```
Comment (erezh31) [inline Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py]: Another small change - add boolean argument to the script, whether to extract keywords or not

File: Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py [hunk 14]
```diff
@@ -260,25 +209,30 @@ def main():
...
         pass
-
-    # evaluate model
-    human_readable, confusion_matrix, report = evaluate_model(train_text_data,
-                                                              train_tag_data,
-                                                              target_accuracy,
-                                                              max_samples_below_threshold,
-                                                              metric)
-
+    train_tag_data = map(lambda x: x[DBOT_TAG_FIELD], data)
```
Comment (erezh31) [inline Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py]: wrap the logic in a separate function, so the main won't be so long

File: Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py [hunk 15]
```diff
@@ -287,19 +241,66 @@ def main():
         human_readable += "\n\nSkip storing model"
     result_entry = {
         'Type': entryTypes['note'],
-        'Contents': {'scores': report, 'confusion_matrix': confusion_matrix.to_dict()},
+        'Contents': {'Threshold': threshold, 'ConfusionMatrixAtThreshold': confusion_matrix,
```
Comment (erezh31) [inline Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py]: add also y_test & y_pred to the contents

File: Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py [hunk 16]
```diff
@@ -287,19 +241,66 @@ def main():
...
+    threshold = float(res[0]['Contents']['threshold'])
+    confusion_matrix = json.loads(res[0]['Contents']['csr_matrix_at_threshold'])
+    metrics_df = json.loads(res[0]['Contents']['metrics_df'])
+    human_readable = res[0]['HumanReadable']
+    human_readable = '\n'.join(['## Results for No Threshold',
+                                'The following results were achieved by using no threshold (threshold equals 0)',
+                                human_readable])
+    result_entry = {
+        'Type': entryTypes['note'],
+        'Contents': {'Threshold': threshold, 'ConfusionMatrixAtThreshold': confusion_matrix,
```
Comment (erezh31) [inline Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py]: add also y_test & y_pred to the contents

File: Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py [hunk 17]
```diff
@@ -287,19 +241,66 @@ def main():
...
+            'DBotPhishingClassifierNoThresh': {
+                'EvaluationScores': metrics_df,
+                'ConfusionMatrix': confusion_matrix
+            }
+        }
+    }
+    demisto.results(result_entry)
+
+
+def get_predictions_for_test_set(train_text_data, train_tag_data):
```
Comment (erezh31) [inline Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py]: move above the main

Comment (lperryp) [reply Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py]: done!

File: Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.yml [hunk 18]
```diff
@@ -35,19 +34,19 @@ args:
...
 - default: false
-  defaultValue: '0.82'
+  defaultValue: '0.8'
   description: The target accuracy, between 0 and 1.
```
Comment (kirbles19) [inline Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.yml]: The target accuracy, between 0 and 1. Default is 0.8.

File: Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.yml [hunk 19]
```diff
@@ -77,6 +76,13 @@ args:
   name: phishingLabels
   required: false
   secret: false
+- default: false
+  defaultValue: '0.8'
+  description: The rario of the training set out of the entire data. Used for model evaluation
```
Comment (kirbles19) [inline Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.yml]: The ratio of the training set to the entire data set, which is used for model evaluation.

File: Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.yml [hunk 20]
```diff
@@ -113,6 +119,17 @@ args:
   - weighted avgm
   required: false
   secret: false
+  deprecated: true
+- default: false
+  defaultValue: 'true'
+  description: Whether to extract keyworks for model.
```
Comment (kirbles19) [inline Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.yml]: Whether to extract keywords for the model. Can be "true" or "false". Default is "true".

**Comment by kirbles19** [comment]
@lperryp @erezh31 @meirwah Done.

**Comment by erezh31** [comment]
@lperryp small thing I think we need to change - rename in the "GetMLModelEvaluation" every time we say "mails" to "incidents" so it will be generic when we will use it in the future for other models - because the logic is pretty much generic.
What do you think?

**Comment by lperryp** [comment]
> @lperryp small thing I think we need to change - rename in the "GetMLModelEvaluation" every time we say "mails" to "incidents" so it will be generic when we will use it in the future for other models - because the logic is pretty much generic.
> What do you think?



> @lperryp small thing I think we need to change - rename in the "GetMLModelEvaluation" every time we say "mails" to "incidents" so it will be generic when we will use it in the future for other models - because the logic is pretty much generic.
> What do you think?

@erezh31 it's a good point
